mobile/lib/modules/album/services/local_album.service.dart

```diff
@@ -0,0 +1,377 @@
+import 'dart:async';
+import 'dart:io';
+import 'package:collection/collection.dart';
+import 'package:immich_mobile/extensions/collection_extensions.dart';
+import 'package:immich_mobile/modules/album/models/album.model.dart';
+import 'package:immich_mobile/extensions/album_extensions.dart';
+import 'package:immich_mobile/modules/backup/models/backup_album.model.dart';
+import 'package:immich_mobile/shared/models/asset.dart';
+import 'package:immich_mobile/shared/models/etag.dart';
+import 'package:immich_mobile/shared/models/exif_info.dart';
+import 'package:immich_mobile/shared/models/store.dart';
+import 'package:immich_mobile/shared/services/hash.service.dart';
+import 'package:immich_mobile/shared/services/sync.service.dart';
+import 'package:immich_mobile/utils/diff.dart';
+import 'package:isar/isar.dart';
+import 'package:logging/logging.dart';
+import 'package:photo_manager/photo_manager.dart';
+
+class LocalAlbumService {
+  Completer<bool> _localCompleter = Completer()..complete(false);
+  final Logger _log = Logger('LocalAlbumService');
+
+  final Isar _db;
+  final HashService _hashService;
+  final SyncService _syncService;
+
+  LocalAlbumService(this._db, this._hashService, this._syncService);
+
+  /// Checks all selected device albums for changes of albums and their assets
+  /// Updates the local database and returns `true` if there were any changes
+  Future<bool> refreshDeviceAlbums() async {
+    if (!_localCompleter.isCompleted) {
+      // guard against concurrent calls
+      _log.info("refreshDeviceAlbums is already in progress");
+      return _localCompleter.future;
+    }
+
+    _localCompleter = Completer();
+    final Stopwatch sw = Stopwatch()..start();
+    bool changes = false;
+    try {
+      final List<AssetPathEntity> onDevice =
+          await PhotoManager.getAssetPathList(
+        filterOption: FilterOptionGroup(containsPathModified: true),
+      );
+      _log.fine("Found ${onDevice.length} device albums");
+      changes = await _syncLocalAlbumAssetsToDb(onDevice);
+      _log.fine("Syncing completed. Changes: $changes");
+    } finally {
+      _localCompleter.complete(changes);
+    }
+    _log.fine("refreshDeviceAlbums took ${sw.elapsedMilliseconds}ms");
+    return changes;
+  }
+
+  /// Syncs all device albums and their assets to the database
+  /// Returns `true` if there were any changes
+  Future<bool> _syncLocalAlbumAssetsToDb(List<AssetPathEntity> onDevice) async {
```

Comment:
Since you moved this from the sync service, it's no longer concurrency safe (the sync service ran this with its internal lock). If a remote sync runs concurrently, the DB state might be garbage.

Reply:
Ah, so the same asset could be updated at the same time from the remote sync and local sync resulting in one of two sync details getting lost? Would it be okay if, for the time being, we make the mutex in the sync service public and use it in the local album service as well?

Reply:
Yes, that would happen. Possibly for a whole batch of assets.

Making the lock public would work fine. Or moving this function back (but I guess you explicitly want to move all handling of local assets here).